### DISCUSSION TITLE
Watch file

### DISCUSSION THREAD

**Discussion by Decodetalkers** [ebkalderon/tower-lsp#366] [Q&A] [answered]
How to watch a file? I found there is a founction `did_change_watched_files`, how to use it?

**Comment by ebkalderon** [comment]
Hello there! According to the [relevant section](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) in the Language Server Protocol spec, the `did_change_watched_files()` notification handler will usually not be called unless explicitly registered by the server after it's already initialized, e.g. inside the `initialized()` notification. Registering for file events looks like this: https://github.com/ebkalderon/tower-lsp/issues/190#issuecomment-647363309

After this has been done, the `did_change_watched_files()` notification will be called by the client as needed, and the specific file events you asked to be watched will be passed in as `params`.

**Reply by Decodetalkers** [reply]
> Hello there! According to the [relevant section](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) in the Language Server Protocol spec, the `did_change_watched_files()` notification handler will usually not be called unless explicitly registered by the server after it's already initialized, e.g. inside the `initialized()` notification. Registering for file events looks like this: [#190 (comment)](https://github.com/ebkalderon/tower-lsp/issues/190#issuecomment-647363309)
> 
> After this has been done, the `did_change_watched_files()` notification will be called by the client as needed, and the specific file events you asked to be watched will be passed in as `params`.

```rust
    async fn initialized(&self, _: InitializedParams) {
        let filechangeparms = DidChangeWatchedFilesRegistrationOptions {
            watchers: vec![FileSystemWatcher {
                glob_pattern: "build/CMakeCache.txt".to_string(),
                kind: Some(lsp_types::WatchKind::all()),
            }],
        };
        let registion = Registration {
            id: "myserver-fileWatcher".to_string(),
            method: "workspace/didChangeWatchedFiles".to_string(),
            register_options: Some(serde_json::to_value(filechangeparms).unwrap()),
        };
        self.client
            .register_capability(vec![registion])
            .await
            .unwrap();
        tracing::info!("init");
        //self.client.register_capabilities(vec![registration]).await.unwrap();
        self.client
            .log_message(MessageType::INFO, "initialized!")
            .await;
    }
  ```

I try to create , edit CMakeCache.txt any where, but nothing changed
Like this? but it seems not work

**Reply by ebkalderon** [reply]
Yes, that looks correct. Have you verified on the the LSP client side that it has successfully received the registration and is detecting the file pattern you specified? It is the responsibility of the client to do the actual file watching and to notify the server as needed, after this point.

**Reply by Decodetalkers** [reply]
> Yes, that looks correct. Have you verified on the the LSP client side that it has successfully received the registration and is detecting the file pattern you specified? It is the responsibility of the client to do the actual file watching and to notify the server as needed, after this point.

No, it not, code is here
https://github.com/Decodetalkers/neocmakelsp/blob/filewatcher/src/main.rs#L115
what ever I  do with that file,  it will not goto this line

**Reply by ebkalderon** [reply]
Again, it sounds like you may need to look at the messages being sent between the client and server and verify that the client is receiving the registration, is watching for the glob pattern you specified, and that the glob pattern is correct (see [the relevant section in the specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles)).

The client is responsible for actually watching files, not the server, so if the dynamic registration wasn't accepted on the client side (or the pattern is somehow incorrect) then the client will never call `did_change_watched_files` on the server.

**Comment by ratmice** [comment, marked as answer]
In the `initialize`'s, `InitializeParams` you might check the `did_change_watched_files` field of `capabilities.workspace`, IIRC this shows whether the client even supports watching files, and there are definitely editors in use that do not.

**Reply by Decodetalkers** [reply]
> In the `initialize`'s, `InitializeParams` you might check the `did_change_watched_files` field of `capabilities.workspace`, IIRC this shows whether the client even supports watching files, and there are definitely editors in use that do not.

It is none, I think I have already set it in initialized..

**Reply by Decodetalkers** [reply]
> In the `initialize`'s, `InitializeParams` you might check the `did_change_watched_files` field of `capabilities.workspace`, IIRC this shows whether the client even supports watching files, and there are definitely editors in use that do not.

https://github.com/neovim/neovim/issues/16078 ohhh, what the.. fuck?

**Reply by ratmice** [reply]
> > In the `initialize`'s, `InitializeParams` you might check the `did_change_watched_files` field of `capabilities.workspace`, IIRC this shows whether the client even supports watching files, and there are definitely editors in use that do not.
> 
> It is none, I think I have already set it in initialized..

It has been a while since I looked at it, but the `None` value here, means the client doesn't support it and so registering won't do anything (at best), it isn't required to be supported by the protocol.

> > In the `initialize`'s, `InitializeParams` you might check the `did_change_watched_files` field of `capabilities.workspace`, IIRC this shows whether the client even supports watching files, and there are definitely editors in use that do not.
> 
> [neovim/neovim#16078](https://github.com/neovim/neovim/issues/16078) ohhh, what the.. fuck?

Yeah, my experience too was that I couldn't get it to work in neovim either.

**Reply by ebkalderon** [reply]
Interesting that Neovim doesn't support this feature yet. In the meantime, are you able to confirm that registering a file watcher works in another editor that supports it? Otherwise, I think this ticket can probably be marked "resolved."
